### PR TITLE
[10.0][MIG] purchase_order_approved

### PR DESCRIPTION
**Pull request by LoisRForgeFlow**
Migration to v10.

Purchase Order Approved
=======================

This module extends the functionality of purchases adding a new state
*Approved* in purchase orders before the *Purchase Order* state. Additionally
add the possibility to set back to draft a purchase order in all the states
previous to *Purchase Order*.

In this new *Approved* state:

* You cannot modify the purchase order.
* However, you can go back to draft and pass through the workflow again.
* The incoming shipments are not created. You can create them by clicking the
  *Convert to Purchase Order* button, also moving to state *Purchase Order*.

### CODE REVIEW

File: purchase_order_approved/models/purchase_order.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    company_id = fields.Many2one(states=READONLY_STATES)
+    picking_type_id = fields.Many2one(states=READONLY_STATES)
+
+    @api.multi
+    def button_release(self):
+        super(PurchaseOrder, self).button_approve()
+
+    @api.multi
+    def button_approve(self):
+        for purchase in self:
```
Comment (rousseldenis) [inline purchase_order_approved/models/purchase_order.py]: @lreficent You can make a write on recordset (if button_approve is called anywhere else than the view) : 

```python
approve_purchases = self.filtered(lambda p: p.company_id.purchase_approve_active)
approve_purchases.write({'state': 'approved'})
return super(PurchaseOrder, self - approve_purchases).button_approve()
```

File: purchase_order_approved/views/purchase_order_view.xml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    <record id="purchase_order_form" model="ir.ui.view">
+        <field name="name">purchase.order.form - purchase_order_approved</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form"/>
+        <field name="arch" type="xml">
+            <button name="button_draft" position="attributes">
+                <attribute name="states">cancel,to approve,approved</attribute>
+            </button>
+            <button name="button_approve" position="after">
+                <button name="button_release" type="object" states='approved'
```
Comment (rousseldenis) [inline purchase_order_approved/views/purchase_order_view.xml:16]: @lreficent Is there any security mechanism (group) here (or is this module goal is just to have one more step) ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Minor comment.

LGTM

File: purchase_order_approved/models/purchase_order.py (hunk 1 above)
Comment (LoisRForgeFlow) [reply purchase_order_approved/models/purchase_order.py]: Sure, thanks for pointing out!

File: purchase_order_approved/views/purchase_order_view.xml (hunk 2 above)
Comment (LoisRForgeFlow) [reply purchase_order_approved/views/purchase_order_view.xml:16]: the goal is just the extra step.

**Review by rousseldenis — APPROVED** [review]

File: purchase_order_approved/README.rst [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+In this new *Approved* state:
+
+* You cannot modify the purchase order.
+* However, you can go back to draft and pass through the workflow again.
+* The incoming shipments are not created. You can create them by clicking the
+  *Convert to Purchase Order* button, also moving to state *Purchase Order*.
+
+The new state diagram is depicted below:
+
+.. image:: purchase_order_approved/static/description/schema.png
```
Comment (HviorForgeFlow) [inline purchase_order_approved/README.rst:23]: fix link to image -> static/description/schema.png

**Review by HviorForgeFlow — APPROVED** [review]
LGTM :+1:

File: purchase_order_approved/models/purchase_order.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+from odoo import api, fields, models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = 'purchase.order'
+
+    state = fields.Selection(selection_add=[('approved', 'Approved')])
+    # TODO: inhterit state but adding approved state in a position after 'to
+    # aprove' state.
```
Comment (gurneyalex) [inline purchase_order_approved/models/purchase_order.py:13]: s/aprove/approve/

File: purchase_order_approved/views/purchase_order_view.xml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2017 Eficent Business and IT Consulting Services S.L.
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+
+<odoo>
+
+    <record id="purchase_order_form" model="ir.ui.view">
```
Comment (gurneyalex) [inline purchase_order_approved/views/purchase_order_view.xml:7]: would be nice to have a filter for "to approve" PO in the search view.

**Review by gurneyalex — APPROVED** [review]

File: purchase_order_approved/views/purchase_order_view.xml (hunk 5 above)
Comment (LoisRForgeFlow) [reply purchase_order_approved/views/purchase_order_view.xml:7]: It is already available. "to approve" state is default in Odoo.
selection_042

**Pull request merged**